### DISCUSSION TITLE
mqtt and dns resolver lib - option to select source interface

### DISCUSSION THREAD

**Discussion by gkujawsk** [zephyrproject-rtos/zephyr#48736] [Ideas] [answered]
In my particular use case I have both eth and ppp which can be optionally selected as the default interface for internet communication. As of now neither mqtt lib nor dns resolver lib does not gives the option to specify the source interface. When both eth and ppp are up and running both with its own default routes it's not clear which of two will be the source interface. 
The default interface set using net_if_set_default() is not taken into the account. 

It would be nice to have the option to specify the source interface or to have both libs using the interface set by net_if_set_default().

In case of mqtt lib it's just a matter of hacking mqtt_transport_*.c so that SO_BINDTODEVICE is set accordingly to the net_if_get_default().  In case of dns resolver it's more tricky. It looks like it's still uses legacy net_context stuff ;-(

**Comment by jukkar** [comment, marked as answer]
About the DNS resolving, this issue is not related to using net_context here. The system selects the output network interface according to destination address. Unfortunately for DNS, the selection might go "wrong" if for example 8.8.8.8 is used as a DNS server and there are multiple network interface that might get selected when deciding what interface to use.
For the normal BSD socket DNS resolving APIs like `gethostbyname()`, there is no way to select the network interface by the user. So the DNS binding to a specific network interface would need to be done by some way before resolving happens.
Selecting a desired network interface could be done for example before this call
https://github.com/zephyrproject-rtos/zephyr/blob/5897de1e535816247e309348cbe29760361a28b8/subsys/net/lib/dns/resolve.c#L296
We could have a call to `net_context_set_iface(context, iface);` just before the bind call, and use desired network interface depending on some external configuration. If you have some ideas how to do this, please propose a PR.

cc: @rlubos

**Reply by gkujawsk** [reply]
Thank you @jukkar for pointing me in the right direction. I'm not proficient enough in the programming skills to propose the PR, but I will simply hack the resolve.c to incorporate the suggested solution.

**Reply by rlubos** [reply]
As for MQTT, I think it should not enforce any particular interface by default, the interface should be selected based on the destination address by the IP layer. I think however, it might be useful to allow MQTT to choose the interface of interest optionally, for example if both interfaces provide a route to an external network.

This could be easily achieved by extending the [mqtt_transport structure](https://github.com/zephyrproject-rtos/zephyr/blob/main/include/zephyr/net/mqtt.h#L392) with an additional pointer to the interface name. NULL pointer could indicate that there's no interface preference, otherwise `SO_BINDTODEVICE` could be used in the MQTT transport implementation to bind to the selected interface.
